pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/CognitiveComplexityBaseVisitor.java

```diff
@@ -0,0 +1,307 @@
+import net.sourceforge.pmd.lang.java.ast.ASTForStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTIfStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTLambdaExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTMethodDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTName;
+import net.sourceforge.pmd.lang.java.ast.ASTPrimaryPrefix;
+import net.sourceforge.pmd.lang.java.ast.ASTStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTSwitchStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTUnaryExpressionNotPlusMinus;
+import net.sourceforge.pmd.lang.java.ast.ASTWhileStatement;
+import net.sourceforge.pmd.lang.java.ast.JavaNode;
+import net.sourceforge.pmd.lang.java.ast.JavaParserVisitorReducedAdapter;
+import net.sourceforge.pmd.lang.java.metrics.impl.visitors.CognitiveComplexityBaseVisitor.State.BooleanOp;
+import net.sourceforge.pmd.lang.java.symboltable.MethodNameDeclaration;
+
+
+/**
+ * Default visitor for the calculation of Cognitive Complexity.
+ *
+ * @see net.sourceforge.pmd.lang.java.metrics.impl.CognitiveComplexityMetric
+ *
+ * @deprecated Visitor decorators are deprecated because they lead to fragile code.
+ */
+@Deprecated
+@InternalApi
+public class CognitiveComplexityBaseVisitor extends JavaParserVisitorReducedAdapter {
+
+    /** Instance. */
+    public static final CognitiveComplexityBaseVisitor INSTANCE = new CognitiveComplexityBaseVisitor();
+
+
+    public static class State {
+        public enum BooleanOp { AND, OR }
+
+        private int complexity = 0;
+        private int nestingLevel = 0;
+
+        private BooleanOp currentBooleanOperation = null;
+        private String methodName = null;
+
+        public double getComplexity() {
+            return complexity;
+        }
+
+        void structureComplexity() {
+            complexity += 1;
+        }
+
+        void nestingComplexity() {
+            complexity += nestingLevel;
+        }
+
+        void booleanOperation(BooleanOp op) {
+            if (currentBooleanOperation != op) {
+                if (op != null) {
+                    structureComplexity();
+                }
+
+                currentBooleanOperation = op;
+            }
+        }
+
+        void increaseNestingLevel() {
+            structureComplexity();
+            nestingComplexity();
+            nestingLevel++;
+        }
+
+        void increaseNestingLevelWithoutStructuralComplexity() {
+            nestingLevel++;
+        }
+
+        void decreaseNestingLevel() {
+            nestingLevel--;
+        }
+
+        void methodCall(String methodCalledName) {
+            if (methodCalledName.equals(methodName)) {
+                structureComplexity();
+            }
+        }
+
+        void setMethodName(String name) {
+            methodName = name;
+        }
+    }
+
+    @Override
+    public Object visit(ASTIfStatement node, Object data) {
+        State state = (State) data;
+
+        for (JavaNode child : node.children()) {
```

Comment:
I think you could avoid looking at children, but instead look at the parent of the current if.

```java
boolean isElseIf = node.getNthParent(2) instanceof ASTIfStatement);
if (!isElseIf) state.increaseNestingLevel();
super.visit(node, data);
if (!isElseIf) state.decreaseNestingLevel();
```

Although if I understand properly, if `isElseIf` you need to pay a hybrid increment, otherwise a structural increment. It looks like it's not the case in your code, or in the sample I've written here. And it looks like it's not tested.